File: packages/gg-sandbox-python/src/gg/tasks.py
"""A private task list, kept as a directed acyclic graph rather than as a list of lines.

Every task may name the tasks that must finish before it, and an edge that would close a cycle is
refused.

One spelling here is worth reading twice. A task's description is a three-way edit, and Python
already spells two of the three: leaving the argument out keeps what is there, `None` clears it, and
a string replaces it. `gg.core.UNCHANGED` is the name of the first, for a program that spells it out.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from wit_world.imports import tasks as wire
from wit_world.imports import types as wire_types

from ._registry import missing, operation
from .core import UNCHANGED, Unchanged, _call, _strings

__all__ = [
    "TaskStatus",
    "TaskUsage",
    "add_task",
    "complete_task",
    "remove_task",
    "set_blocked_by",
    "update_task",
]


class TaskStatus(Enum):
    """Where a task stands."""

    PENDING = "pending"
    """Not started. Every task begins here."""

    IN_PROGRESS = "in_progress"
    """Being worked on now."""

    DONE = "done"
    """Finished. Tasks blocked on it become actionable once all their blockers are done."""


@dataclass(frozen=True)
class TaskUsage:
    """How much of the run's task budget is used, after the call that returned it."""

    count: int
    """Tasks currently on the list."""

    max_tasks: int
    """The most tasks this run allows."""


def _text_edit(value: str | None | Unchanged) -> wire_types.TextEdit:
    """Lower the `UNCHANGED` / `None` / string sentinel onto the membrane's three-way text edit."""
    if value is UNCHANGED:
        return wire_types.TextEdit_Keep()
    if value is None:
        return wire_types.TextEdit_Clear()
    return wire_types.TextEdit_Set(value)


def _status(status: TaskStatus | None) -> wire.TaskStatus | None:
    """Lower a model-facing status onto the membrane's, whose `in-progress` carries a hyphen."""
    return None if status is None else wire.TaskStatus[status.name]


def _usage(usage: wire.TaskUsage) -> TaskUsage:
    """The membrane's budget record, as the model-facing one."""
    return TaskUsage(count=usage.count, max_tasks=usage.max_tasks)


@operation("tasks.add_task")
def add_task(
    id: str,
    title: str,
    *,
    description: str | None = None,
    blocked_by: list[str] | None = None,
) -> TaskUsage:
    """Add a task to the task graph.

    `blocked_by` names the tasks that must finish before this one, and defaults to none.

    Args:
        id: The id chosen for it. Every other task call takes it, and no two tasks may share one.
        title: A short line naming the work.
        description: What the work is, at whatever length is useful.
        blocked_by: The ids of the tasks that must be done before this one. The default is none.

    Returns:
        The task budget the addition left behind.

    Raises:
        ToolError: `invalid-argument` for a blank id or title, `conflict` on a duplicate id or an
            edge that would close a cycle, `not-found` for an unknown blocker, and `limit-exceeded`
            at the task cap.
    """
    return _usage(
        _call(
            wire.add_task,
            wire.TaskInput(
                id=id,
                title=title,
                description=description,
                blocked_by=_strings("add_task", "blocked_by", blocked_by),
            ),
        )
    )


@operation("tasks.update_task")
def update_task(
    id: str,
    *,
    title: str | None = None,
    description: str | None | Unchanged = UNCHANGED,
    status: TaskStatus | None = None,
) -> None:
    """Revise a task's title, description or status; at least one must be supplied.

    An argument left out is left alone. `description=None` empties the description, which is why it
    defaults to `UNCHANGED` rather than to `None`.

    Args:
        id: The task to revise.
        title: The title to replace the old one with. The default keeps the one it has.
        description: The description to replace the old one with; `None` clears it, and the default
            keeps the one it has.
        status: Where the task now stands. The default keeps the status it has.

    Raises:
        ToolError: `invalid-argument` when no field was supplied or one was blanked, and `not-found`
            for an unknown id.
    """
    _call(
        wire.update_task,
        id,
        wire.TaskPatch(
            title=title,
            description=_text_edit(description),
            status=_status(status),
        ),
    )


@operation("tasks.set_blocked_by")
def set_blocked_by(id: str, blocked_by: list[str]) -> None:
    """Replace a task's whole blocker set; an empty list clears every blocker.

    Args:
        id: The task whose blockers to replace.
        blocked_by: The ids of every task that must now be done before it. An empty list clears them
            all.

    Raises:
        ToolError: `invalid-argument` for a blank id or blocker, `not-found` for a task or blocker
            that is not on the list, and `conflict` when an edge would close a cycle or block the
            task on itself.
    """
    _call(wire.set_blocked_by, id, _strings("set_blocked_by", "blocked_by", blocked_by))


@operation("tasks.complete_task")
def complete_task(id: str) -> None:
    """Mark a task done.

    Tasks it was blocking become actionable once every one of their blockers is done.

    Args:
        id: The task to mark done.

    Raises:
        ToolError: `not-found` for an unknown id.
    """
    _call(wire.complete_task, id)


@operation("tasks.remove_task")
def remove_task(id: str) -> TaskUsage:
    """Remove a task and every blocker edge pointing at it.

    Args:
        id: The task to remove.

    Returns:
        The task budget the removal left behind.

    Raises:
        ToolError: `not-found` for an unknown id.
    """
    return _usage(_call(wire.remove_task, id))


__getattr__ = missing(__name__, __all__)
"""What this module answers for a name it does not declare — see `gg._registry.missing`."""

File: packages/gg-sandbox-ruby/src/gg/tasks.rb
# frozen_string_literal: true

module GG
  # A private task list, kept as a directed acyclic graph rather than as a list of lines.
  #
  # Every task may name the tasks that must finish before it, and an edge that would close a cycle
  # is refused.
  #
  # A description is a three-way edit: leaving the argument out, or passing `GG::Core::UNCHANGED`,
  # keeps it; `nil` clears it; a string replaces it.
  module Tasks
    extend Surface::Operations

    # The membrane's budget record, as the model-facing one.
    #
    # @param usage [Object] the wire record
    # @return [GG::Tasks::TaskUsage] the budget a program reads
    # @api private
    def self.usage(usage)
      TaskUsage.new(
        count: Wire.field(usage, "count"),
        max_tasks: Wire.field(usage, "maxTasks")
      )
    end
    private_class_method :usage

    # Add a task to the task list and hand back the task budget.
    #
    # @param id [String] The id this task takes. Every other task call takes it, and no two tasks
    #   may share one.
    # @param title [String] A short line naming the work.
    # @param description [String, nil] What the work is, at whatever length is useful.
    # @param blocked_by [Array<String>] The ids of the tasks that must be done before this one.
    #   Defaults to none.
    # @return [GG::Tasks::TaskUsage] how much of the task budget is now used
    # @raise [GG::Core::ApiError] `:conflict` on a duplicate id or on an edge that would close a
    #   cycle.
    def self.add_task(id, title, description: nil, blocked_by: [])
      usage(Wire.call("add_task", "tasks", "addTask", [
                        Wire.record(
                          "id" => id,
                          "title" => title,
                          "description" => description,
                          "blockedBy" => Check.strings("add_task", "blocked_by", blocked_by)
                        )
                      ]))
    end
    operation :add_task, "tasks.add_task", tool: "add_task"

    # Revise a task's title, description or status; at least one of them is required.
    #
    # @param id [String] The task to revise.
    # @param title [String, nil] The title to replace the old one with. Leave it out to keep the one
    #   it has.
    # @param description [String, nil, GG::Core::Unchanged] The description to replace the old one
    #   with; `nil` clears it, and leaving it out keeps the one it has.
    # @param status [GG::Tasks::TaskStatus, nil] Where the task now stands. Leave it out to keep the
    #   status it has.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:not_found` for an unknown id.
    def self.update_task(id, title: nil, description: Core::UNCHANGED, status: nil)
      Wire.call("update_task", "tasks", "updateTask", [
                  id,
                  Wire.record(
                    "title" => title,
                    "description" => Wire.text_edit(description),
                    "status" => Wire.arm(Check.choice("update_task", "status", status,
                                                      TaskStatus::ALL))
                  )
                ])
      nil
    end
    operation :update_task, "tasks.update_task", tool: "update_task"

    # Replace a task's whole blocker set; passing no blockers clears every one of them.
    #
    # @param id [String] The task whose blockers to replace.
    # @param blocked_by [Array<String>] The ids of every task that must now be done before it,
    #   splatted. Passing none clears them all.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:not_found` for an unknown id, and `:conflict` when an edge
    #   would close a cycle.
    def self.set_blocked_by(id, *blocked_by)
      Wire.call("set_blocked_by", "tasks", "setBlockedBy",
                [id, Check.strings("set_blocked_by", "blocked_by", blocked_by)])
      nil
    end
    operation :set_blocked_by, "tasks.set_blocked_by", tool: "set_blocked_by"

    # Mark a task done.
    #
    # Tasks it was blocking become actionable once every one of their blockers is done.
    #
    # @param id [String] The task to mark done.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:not_found` for an unknown id.
    def self.complete_task(id)
      Wire.call("complete_task", "tasks", "completeTask", [id])
      nil
    end
    operation :complete_task, "tasks.complete_task", tool: "complete_task"

    # Remove a task and every blocker edge pointing at it, and hand back the task budget.
    #
    # @param id [String] The task to remove.
    # @return [GG::Tasks::TaskUsage] how much of the task budget is still used
    # @raise [GG::Core::ApiError] `:not_found` for an unknown id.
    def self.remove_task(id)
      usage(Wire.call("remove_task", "tasks", "removeTask", [id]))
    end
    operation :remove_task, "tasks.remove_task", tool: "remove_task"

    # Where a task stands.
    #
    # Every arm is a Symbol: `GG::Tasks::TaskStatus::DONE` and `:done` are the same value.
    module TaskStatus
      # Not started. Every task begins here.
      PENDING = :pending

      # Being worked on now.
      IN_PROGRESS = :in_progress

      # Finished. Tasks blocked on it become actionable once all their blockers are done.
      DONE = :done

      # Every status a task may be moved to.
      ALL = [PENDING, IN_PROGRESS, DONE].freeze
    end

    # How much of the run's task budget is used, after the call that returned it.
    class TaskUsage
      include Value

      # @return [Integer] Tasks currently on the list.
      attr_reader :count

      # @return [Integer] The most tasks this run allows.
      attr_reader :max_tasks

      # @api private
      def initialize(count:, max_tasks:)
        @count = count
        @max_tasks = max_tasks
        freeze
      end
    end
  end
end
